### PR TITLE
Add the Duktape stack trace to the thrown DuktapeException as StackTr…

### PR DESCRIPTION
**Pull request by szurbrigg**
…aceElements rather than including it in the message string.

### CODE REVIEW

File: duktape/src/main/java/com/squareup/duktape/DuktapeException.java [hunk 1]
```diff
@@ -15,8 +15,58 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
 public final class DuktapeException extends RuntimeException {
+  /**
+   * Duktape stack trace strings have multiple lines of the format "function filename.ext:line".
+   * "function" is optional, but we'll omit those frames from the exception, since it means the
```
Comment (swankjesse) [inline duktape/src/main/java/com/squareup/duktape/DuktapeException.java]: ... but we'll omit frames without a function...

File: duktape/src/main/java/com/squareup/duktape/DuktapeException.java [hunk 2]
```diff
@@ -15,8 +15,58 @@
...
+   * frame is in native code.
+   */
+  private final static Pattern STACK_TRACE_PATTERN =
+      Pattern.compile("\\s*([^\\s]+) ([^\\s]+):(\\d+).*");
```
Comment (swankjesse) [inline duktape/src/main/java/com/squareup/duktape/DuktapeException.java]: should this have a trailing `$` so the line is at the end? I'm wondering about hypothetical filenames like `hello:123.js`

File: duktape/src/main/java/com/squareup/duktape/DuktapeException.java [hunk 3]
```diff
@@ -15,8 +15,58 @@
...
+  /** Java StackTraceElements require a class name.  We don't have one in JS, so use this. */
+  private final static String STACK_TRACE_CLASS_NAME = "Javascript";
```
Comment (swankjesse) [inline duktape/src/main/java/com/squareup/duktape/DuktapeException.java]: nit: `JavaScript` ?

File: duktape/src/main/java/com/squareup/duktape/DuktapeException.java [hunk 4]
```diff
@@ -15,8 +15,58 @@
...
+
   public DuktapeException(String detailMessage) {
-    super(detailMessage);
+    super(getErrorMessage(detailMessage));
+    String[] lines = TextUtils.split(detailMessage, "\n");
```
Comment (swankjesse) [inline duktape/src/main/java/com/squareup/duktape/DuktapeException.java]: I don't know what `TextUtils` is, but you probably don't want it. It means you can't test on the desktop VM.

Comment (swankjesse) [inline duktape/src/main/java/com/squareup/duktape/DuktapeException.java]: (Probably easier is `detailMessage.split("\n", -1)`)

File: duktape/src/main/java/com/squareup/duktape/DuktapeException.java [hunk 5]
```diff
@@ -15,8 +15,58 @@
...
+   * Duktape stack trace strings have multiple lines of the format "function filename.ext:line".
+   * "function" is optional, but we'll omit those frames from the exception, since it means the
+   * frame is in native code.
+   */
+  private final static Pattern STACK_TRACE_PATTERN =
+      Pattern.compile("\\s*([^\\s]+) ([^\\s]+):(\\d+).*");
+  /** Java StackTraceElements require a class name.  We don't have one in JS, so use this. */
+  private final static String STACK_TRACE_CLASS_NAME = "Javascript";
+
   public DuktapeException(String detailMessage) {
```
Comment (swankjesse) [inline duktape/src/main/java/com/squareup/duktape/DuktapeException.java]: Can you just change this to pass two strings, the error message and the stacktrace? That way if the error message has a `\n` in it, we don't corrupt.

File: duktape/src/main/java/com/squareup/duktape/DuktapeException.java [hunk 6]
```diff
@@ -15,8 +15,58 @@
...
+  }
+
+  private static String getErrorMessage(String detailMessage) {
+    int end = detailMessage.indexOf('\n');
+    return end > 0
+        ? detailMessage.substring(0, end)
+        : detailMessage;
+  }
+
+  private static StackTraceElement toStackTraceElement(String s) {
```
Comment (swankjesse) [inline duktape/src/main/java/com/squareup/duktape/DuktapeException.java]: Awesome.

**Comment by swankjesse** [comment]
This is going to be great. Lots of feedback & I'd like another round once you've addressed this feedback.

File: duktape/src/main/java/com/squareup/duktape/DuktapeException.java (hunk 5 above)
Comment (szurbrigg) [inline duktape/src/main/java/com/squareup/duktape/DuktapeException.java]: I could, but we get that combined string from Duktape, so we'd still have to split on `\n` in JNI.  Also, we'd lose the ability to use the `ThrowNew` shorthand function, so in JNI we'd have to look up the constructor method for the exception, allocate it, and then throw.

IMO, the way it is here is preferable, but I can change it if you feel strongly.

File: duktape/src/main/java/com/squareup/duktape/DuktapeException.java (hunk 4 above)
Comment (szurbrigg) [inline duktape/src/main/java/com/squareup/duktape/DuktapeException.java]: Good point. Will do.

File: duktape/src/main/java/com/squareup/duktape/DuktapeException.java (hunk 3 above)
Comment (szurbrigg) [inline duktape/src/main/java/com/squareup/duktape/DuktapeException.java]: Sure.

File: duktape/src/main/java/com/squareup/duktape/DuktapeException.java (hunk 2 above)
Comment (szurbrigg) [inline duktape/src/main/java/com/squareup/duktape/DuktapeException.java]: Yeah, that makes sense.  (Although, are colons in filenames legal?)

File: duktape/src/main/java/com/squareup/duktape/DuktapeException.java (hunk 1 above)
Comment (szurbrigg) [inline duktape/src/main/java/com/squareup/duktape/DuktapeException.java]: Will change.

File: duktape/src/main/java/com/squareup/duktape/DuktapeException.java (hunk 2 above)
Comment (swankjesse) [inline duktape/src/main/java/com/squareup/duktape/DuktapeException.java]: Everything is legal. Some things are escaped.

File: duktape/src/main/java/com/squareup/duktape/DuktapeException.java (hunk 5 above)
Comment (swankjesse) [inline duktape/src/main/java/com/squareup/duktape/DuktapeException.java]: Nope, if we have to manually split we might as well do it in managed code.

File: duktape/src/main/java/com/squareup/duktape/DuktapeException.java [hunk 7]
```diff
@@ -15,8 +15,56 @@
...
+  private final static Pattern STACK_TRACE_PATTERN =
+      Pattern.compile("\\s*([^\\s]+) ([^\\s]+):(\\d+).*$");
+  /** Java StackTraceElements require a class name.  We don't have one in JS, so use this. */
+  private final static String STACK_TRACE_CLASS_NAME = "JavaScript";
+
   public DuktapeException(String detailMessage) {
-    super(detailMessage);
+    super(getErrorMessage(detailMessage));
+    String[] lines = detailMessage.split("\n", -1);
+    if (lines.length <= 1) {
```
Comment (swankjesse) [inline duktape/src/main/java/com/squareup/duktape/DuktapeException.java]: ultranit: split never returns empty arrays, so this _could_ be `if (lines.length == 1)`

Comment (swankjesse) [inline duktape/src/main/java/com/squareup/duktape/DuktapeException.java]: (Actually, `split()` can return an empty array, but only in broken mode where you don't pass `-1` as the 2nd arg)

**Pull request merged**